sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala

```diff
@@ -283,6 +287,14 @@ abstract class ProgressContext(
     progressReporter.lastNoExecutionProgressEventTime = triggerClock.getTimeMillis()
     progressReporter.updateProgress(newProgress)
 
+    // Ask the state store coordinator to look for any lagging instances and report them.
+    progressReporter.stateStoreCoordinator
+      .constructLaggingInstanceReport(
+        lastExecution.runId,
+        lastEpochId,
```

Comment:
fyi this is a bug. lastEpochId (aka batchId) != latest store version. latest store version = batchId + 1. You can confirm. Because when batch x commits, it produces store version x + 1.

Your test is not catching this because you're doing >= 0 verification in your test. Also because you don't have a test for when changelog is disabled.

Reply:
Hmm gotcha, that's a good catch. Thanks!